src/main/java/com/conveyal/gtfs/validator/FlexValidator.java

```diff
@@ -0,0 +1,358 @@
+            validateLocations(location, Lists.newArrayList(feed.stops), Lists.newArrayList(feed.fareRules))
+        );
+        feed.stopTimes.forEach(stopTime ->
+            validateStopTimes(stopTime, Lists.newArrayList(feed.locationGroups), Lists.newArrayList(feed.locations))
+        );
+        feed.bookingRules.forEach(this::validateBookingRules);
+        // Register errors, if any, once all checks have been completed.
+        errors.forEach(this::registerError);
+    }
+
+    public void validateLocationGroups(LocationGroup locationGroup, List<Stop> stops, List<Location> locations) {
+        if (!stops.isEmpty() && stops.stream().anyMatch(stop -> stop.stop_id.equals(locationGroup.location_group_id))) {
+            // Location group id must not match a stop id.
+            errors.add(NewGTFSError.forEntity(
+                locationGroup,
+                NewGTFSErrorType.FLEX_FORBIDDEN_LOCATION_GROUP_ID).setBadValue(locationGroup.location_group_id)
+            );
+        }
+        if (!locations.isEmpty() &&
+            locations.stream().anyMatch(location -> location.location_id.equals(locationGroup.location_group_id))) {
+            // Location group id must not match a location id.
+            errors.add(NewGTFSError.forEntity(
+                locationGroup,
+                NewGTFSErrorType.FLEX_FORBIDDEN_LOCATION_GROUP_ID).setBadValue(locationGroup.location_group_id)
+            );
+        }
+    }
+
+    public void validateLocations(Location location, List<Stop> stops, List<FareRule> fareRules) {
+        if (!stops.isEmpty() && stops.stream().anyMatch(stop -> stop.stop_id.equals(location.location_id))
+        ) {
+            // Location id must not match a stop id.
+            errors.add(NewGTFSError.forEntity(
+                location,
+                NewGTFSErrorType.FLEX_FORBIDDEN_LOCATION_ID).setBadValue(location.location_id)
+            );
+        }
+        if (fareRules != null &&
+            !fareRules.isEmpty() &&
+            fareRules.stream().anyMatch(fareRule ->
+                (fareRule.contains_id != null && fareRule.destination_id != null && fareRule.origin_id != null) &&
+                !fareRule.contains_id.equals(location.zone_id) &&
+                !fareRule.destination_id.equals(location.zone_id) &&
+                !fareRule.origin_id.equals(location.zone_id)
+            )
+        ) {
+            // zone id is required if fare rules are defined.
+            errors.add(NewGTFSError.forEntity(
+                location,
+                NewGTFSErrorType.FLEX_MISSING_FARE_RULE).setBadValue(location.zone_id)
+            );
+        }
+    }
+
+    public void validateStopTimes(
+        StopTime stopTime,
+        List<LocationGroup> locationGroups,
+        List<Location> locations) {
+
+        if (stopTime.arrival_time != INT_MISSING &&
+            (stopTime.start_pickup_dropoff_window != INT_MISSING ||
+                stopTime.end_pickup_dropoff_window != INT_MISSING)
```

Comment:
Consider extracting a variable or method to replace the comments inside the if block.

Reply:
This might bulk the code unnecessarily if I'm to do it for all similar `if` statements, so I'm in two minds. I think the comments are useful and some of the detail will be lost if converted to a variable. I guess if I went with methods I could put the comment in the method comment, but I'm then back to my original point. If you think this is a show stopper, I'm happy to make the changes, if not, I'm leaning towards leaving it.